Move time source from db layer to PersistenceManager

ribaraka opened the pull request:
<!-- Describe what has changed in this PR -->
Refactored the system so that the PersistenceManager level handles the timestamp;
Removing the timeSrc logic from the DB layer;
The DB layer only translate entity records into database queries without modifying or adding any additional information like timestamps;

<!-- Tell your future self why have you made these changes -->
**Why?**
#6610 


<!-- How have you verified this change? Tested locally? Added a unit test? Checked in staging env? -->
**How did you test it?**
unit tests
integration tests
smoke tests


<!-- Assuming the worst case, what can be broken when deploying this change to production? -->
**Potential risks**

<!-- Is it notable for release? e.g. schema updates, configuration or data migration required? If so, please mention it, and also update CHANGELOG.md -->
**Release notes**

<!-- Is there any documentation updates should be made for config, https://cadenceworkflow.io/docs/operation-guide/setup/ ? If so, please open an PR in https://github.com/cadence-workflow/cadence-docs -->
**Documentation Changes**

bowenxia commented on common/persistence/nosql/nosql_execution_store_util_test.go:
```diff
@@ -60,7 +60,7 @@ func TestNosqlExecutionStoreUtils(t *testing.T) {
 					},
 				}
-				return store.prepareCreateWorkflowExecutionRequestWithMaps(workflowSnapshot)
+				return store.prepareCreateWorkflowExecutionRequestWithMaps(workflowSnapshot, time.Now())
```
Have you had any error when running the unit tests? I was thinking to use a constant time stamp in the tests like [this ](https://github.com/cadence-workflow/cadence/blob/e57792ddb62cf10fd51bb3c13606da33fea12c72/common/persistence/nosql/nosqlplugin/cassandra/workflow_utils_test.go#L45) so that in the unit tests it will always keep the same

ribaraka replied:
No, there haven't been any errors in the tests with this implementation. However, I refactored it to use a FixedTime variable for consistency.

ribaraka commented:
@bowenxia added another commit with a refactor of the history storage. If you have time, please take a look. I'll be sending the work in small chunks, little by little. There are a few TODOs— if you could reply to them, I’d really appreciate it.

bowenxia commented on common/persistence/nosql/nosqlplugin/cassandra/history_events.go:
```diff
@@ -54,19 +53,20 @@ func (db *cdb) InsertIntoHistoryTreeAndNode(ctx context.Context, treeRow *nosqlp
 		batch := db.session.NewBatch(gocql.LoggedBatch).WithContext(ctx)
 		batch.Query(v2templateInsertTree,
-			treeRow.TreeID, treeRow.BranchID, ancs, persistence.UnixNanoToDBTimestamp(treeRow.CreateTimestamp.UnixNano()), treeRow.Info, timeStamp)
+			// TODO: Should the `persistence.UnixNanoToDBTimestamp(treeRow.CreateTimestamp.UnixNano())` logic be handled by `persistenceManager` as well, or should it remain as is?
```
Let's leave it as is, since this PR is more related to the created_time and last_updated_time timestamps.

bowenxia commented:
LGTM so far. Lemme know if you are planning to refactor other persistence managers here, or in a new PR.

ribaraka commented:
> LGTM so far. Lemme know if you are planning to refactor other persistence managers here, or in a new PR.

I'm planning to do everything in one PR. I figured it would be more convenient to have the refactor in a single place rather than splitting it up, unless there are any objections.

ribaraka commented:
@bowenxia, I refactored the task manager—please take a look. I've added some questions marked as TODOs.

bowenxia commented on common/persistence/nosql/nosqlplugin/cassandra/tasks.go:
```diff
@@ -387,23 +387,25 @@ PopulateTasks:
-				ttl = &ttlInt // TTL is an integer
-			}
-		}
+		// TODO: no usage of ttl
```
Let's leave it as-is. Because this is not the purpose of this PR. 
If there's something wrong, it would be easier to find root cause.

bowenxia commented on common/persistence/nosql/nosqlplugin/cassandra/tasks.go:
```diff
@@ -387,23 +387,25 @@ PopulateTasks:
-		if ttl != nil {
-			t.Expiry = db.timeSrc.Now().Add(time.Duration(*ttl) * time.Second)
-		}
+		// TODO: removing this, because there is no usage of it
```
same here

ribaraka replied:
Another issue at this low-level layer is that TTL expiry **computation** is happening here, whereas I understand that we should remove any **computation** from db layer entirely. Moving this logic to upstream led nowhere, as TTL isn't used there either.

Since the task is to remove `timeSrc` from `db` and we need to discuss this matter later, I could replace the current logic with `time.Now()` and keep the TODO.

bowenxia commented on common/persistence/nosql/nosqlplugin/cassandra/tasks.go:
```diff
@@ -110,7 +110,7 @@ func fromTaskListPartitionConfig(config *persistence.TaskListPartitionConfig) ma
 // Return TaskOperationConditionFailure if the condition doesn't meet
 func (db *cdb) InsertTaskList(ctx context.Context, row *nosqlplugin.TaskListRow) error {
-	timeStamp := db.timeSrc.Now()
+	timeStamp := row.LastUpdatedTime
```
insert is a create operation. It would be confused to have lastUpdatedTime. I understand that you might want to reuse the LastUpdateTime in TaskListRow struct, but maybe rename it to like "timestamp" would be a good idea.

ribaraka replied:
I suggest using a single expression, like `currentTimestamp`, for all functions (insert, update). What do you think?

bowenxia commented on common/persistence/nosql/nosqlplugin/cassandra/tasks.go:
```diff
@@ -282,7 +282,7 @@ func (db *cdb) InsertTasks(
 	taskListName := tasklistCondition.TaskListName
 	taskListType := tasklistCondition.TaskListType
-	timeStamp := db.timeSrc.Now()
+	timeStamp := tasklistCondition.LastUpdatedTime
```
same here. Insert is a create operation

ribaraka commented:
@bowenxia when you have a moment, please take another look—especially at the Queue Manager. It requires additional attention on how to handle it.

bowenxia commented on common/persistence/nosql/nosql_queue_store.go:
```diff
@@ -114,16 +115,11 @@ func (q *nosqlQueueStore) EnqueueMessageToDLQ(
 	}
 
-	_, err = q.tryEnqueue(ctx, q.getDLQTypeFromQueueType(), lastMessageID+1, messagePayload)
+	_, err = q.tryEnqueue(ctx, q.getDLQTypeFromQueueType(), lastMessageID+1, messagePayload, time.Now())
```
Let's do the same as you did for UpdateAckLevel and UpdateDLQAckLevel to have a currentTimestamp as an extra parameter

ribaraka commented:
@bowenxia, Could you please review the shard refactoring? Is it okay to modify the ShardCRUD interface so far?

bowenxia commented on common/persistence/nosql/nosql_task_store.go:
```diff
@@ -115,14 +115,14 @@ func (t *nosqlTaskStore) LeaseTaskList(
-				RangeID:         initialRangeID,
-				TaskListKind:    request.TaskListKind,
-				AckLevel:        initialAckLevel,
-				LastUpdatedTime: now,
```
nit: consider change `now` to be a more descriptive name like `currentTimeStamp`

bowenxia commented on common/persistence/nosql/nosql_task_store.go:
```diff
@@ -329,12 +329,12 @@ func (t *nosqlTaskStore) CreateTasks(
-		TaskListName:    request.TaskListInfo.Name,
-		TaskListType:    request.TaskListInfo.TaskType,
-		RangeID:         request.TaskListInfo.RangeID,
-		LastUpdatedTime: now,
```
nit: same

bowenxia reviewed (commented):
shard storage refactored LGTM

bowenxia commented on common/persistence/nosql/nosql_execution_store.go:
```diff
@@ -830,13 +830,14 @@ func (d *nosqlExecutionStore) CreateFailoverMarkerTasks(
 		if err != nil {
 			return err
 		}
+		tasks[i].CurrentTimeStamp = request.CurrentTimeStamp
```
It is redundant here to add timestamp to each task since we are storing the entire tasklist into Cassandra; also we do have a creationTime for each task. 

I am thinking how to pass one timestamp into the method `d.db.InsertReplicationTask` without changing the signature. @taylanisikdemir Any suggestions? I am thinking to add the timestamp into ctx, but not sure if it is the best way.

ribaraka replied:
There has been update in this block of code (i.e. Refactor peristence layer to support inserting history tasks of new categories #6671). Adjusted the code accordingly.

ribaraka replied:
There have been another new changes and this block of code has been restored.

bowenxia commented on common/persistence/nosql/nosql_queue_store.go:
```diff
@@ -100,7 +101,7 @@ func (q *nosqlQueueStore) EnqueueMessage(
 		return err
 	}
-	_, err = q.tryEnqueue(ctx, q.queueType, getNextID(ackLevels, lastMessageID), messagePayload)
+	_, err = q.tryEnqueue(ctx, q.queueType, getNextID(ackLevels, lastMessageID), messagePayload, time.Now())
```
Yes you are right. 

Adding timeSrc to nosql_queue_store might be a way. @taylanisikdemir Do you have an opinion? Since it is too hard to add timeSrc in queue_manager and pass it down to nosql_queue_store.

bowenxia replied:
Update: 

Sync with Taylan offilne. Let's add timeSrc as a field in to nosql_queue_store, and then create it during the call of its constructor for now. Also, could you mark this part as a to do for future work? Thanks!

bowenxia commented:
Reviewing new changes

ribaraka commented:
@bowenxia, I’ve made some minor changes and added one TODO—please take a look.

Looks like I’ve refactored everything required for this task. The Queue store is still a bit tricky though.
I’ve adjusted all related tests accordingly. Unit, integration, and E2E tests are all passing successfully. I also ran a smoke test.

I think it's ready to remove WIP and start to wrap it up :smile:

bowenxia commented:
Hi @ribaraka sounds good. I was too busy to review the PR today. I'll try get back to you tmr or the day after. Thanks for your contribution!

bowenxia commented:
LGTM, let me ask my teammates if they have any concerns

bowenxia commented:
I've accept the PR, but let's wait until EoW to merge it, just in order to see if there are concerns.

bowenxia commented:
Looks like integration test failed. Could you take another look to fix those please?

ribaraka commented:
>  Looks like integration test failed. Could you take another look to fix those please?



@bowenxia I can't pinpoint the exact cause of these failures. I ran the integration tests locally, and everything seems to be working fine. Could you help me identify what's going wrong?

bowenxia commented:
Retried several times and it finally worked. Feel free to merge @ribaraka :)

ribaraka commented:
@bowenxia, thank you for all the support you've given me so far. One more thing : 

> Feel free to merge @ribaraka :)

I'm not able to do that. Should I ask someone?
